DOC: update the NumPy Roadmap

rgommers opened the pull request:
With NumPy 2.0 RCs available, that release is feature-complete and a lot on the roadmap is outdated. So time for a large update.

rgommers committed:
DOC: update the NumPy Roadmap

With NumPy 2.0 RCs available, that release is feature-complete
and a lot on the roadmap is outdated. So time for a large update.

[skip actions] [skip cirrus] [skip azp]

charris commented on doc/neps/roadmap.rst:
```diff
@@ -68,69 +70,115 @@ documentation on many topics are missing or outdated. See :ref:`NEP44` for
+
+We will need to assess the costs and benefits to NumPy itself,
+downstream package authors, and end users. Based on that assessment
+
```
Sentence fragment, "Based on that assessment ...".

charris commented on doc/neps/roadmap.rst:
```diff
@@ -68,69 +70,115 @@ documentation on many topics are missing or outdated. See :ref:`NEP44` for
-- A backend system for ``numpy.fft`` (so that e.g. ``fft-mkl`` doesn't need to monkeypatch numpy).
 - Write a strategy on how to deal with overlap between NumPy and SciPy for ``linalg``.
-- Deprecate ``np.matrix`` (very slowly).
+- Deprecate ``np.matrix`` (very slowly) - this is feasible ones the switch-over
```
ones -> once

charris commented on doc/neps/roadmap.rst:
```diff
@@ -68,69 +70,115 @@ documentation on many topics are missing or outdated. See :ref:`NEP44` for
+well in NumPy. After that is stable and complete, there may be opportunities to
+actually make use of the potential for performance improvements from
+free-threaded CPython, or make it easier to do so for NumPy's users.
+
```
Releasing free-threaded wheels is going to significantly increase the number of wheels on PyPI as we drop Python versions that don't support it. I would hope that at some point we will only need to build for a single Python version.

rgommers replied:
Re number of wheels: that should be a temporary situation. In the discussions around PEP 703, this load on package authors was prominently discussed. The end state should be that the free-threaded becomes the default, and then the "with GIL" build fades away.

charris commented on doc/neps/roadmap.rst:
```diff
@@ -18,25 +18,19 @@ may include (among other things) interoperability protocols, better duck typing
-implement a NumPy-compatible API. This may include defining consistent subsets
-of the API, as discussed in `this section of NEP 37
-<https://numpy.org/neps/nep-0037-array-module.html#requesting-restricted-subsets-of-numpy-s-api>`__.
+In addition, the ``__array_ufunc__`` and ``__array_function__`` protocols
```
My thought at the time those went in was that we needed to do something quickly to slow the fragmentation of the ecosystem and demonstrate the utility of interoperability. Standardization was the long term solution, but I figured it would take 5 - 10 years, and here we are, six years later.

eendebakpt commented on doc/neps/roadmap.rst:
```diff
@@ -68,69 +70,115 @@ documentation on many topics are missing or outdated. See :ref:`NEP44` for
+
+We made a very large amount of changes (and improvements!) in NumPy 2.0. The
+release process has taken a very long time, and part of the ecosystem is still
+catching up. We may need to slow down for a while, and possible help the rest
```
```suggestion
catching up. We may need to slow down for a while, and possibly help the rest
```

aragilar commented on doc/neps/roadmap.rst:
```diff
@@ -68,69 +70,115 @@ documentation on many topics are missing or outdated. See :ref:`NEP44` for
-Concrete goals for the dtype system rewrite are:
-
-- Easier custom dtypes:
+We aim to continue making it easier to extend NumPy. The primary topic here is to
+improve the dtype system - see for example :ref:`NEP41` and related NEPs linked
+from it. In NumPy 2.0, a new C API for user-defined dtypes was made public. We aim
+to encourage its usage and improve this API further.
```
Is the new C API documented anywhere yet? It would be good to put in a link to it if possible.

rgommers replied:
Good point. Docs are at https://numpy.org/devdocs/reference/c-api/array.html#custom-data-types

rgommers committed:
DOC: update for review comments

[skip actions] [skip cirrus] [skip azp]

lorentzenchr commented on doc/neps/roadmap.rst:
```diff
@@ -46,17 +40,25 @@ Improvements to NumPy's performance are important to many users. We have
-SIMD support across all relevant NumPy functions.
+SIMD support across relevant NumPy functionality.
+
+Transitioning from C to C++, both in the SIMD infrastructure and in NumPy
```
Are the main drivers for this choice written somewhere (might help discussion in other packages).

rgommers replied:
https://numpy.org/neps/nep-0054-simd-cpp-highway.html#motivation-and-scope contains some reasons. Some are specific to the SIMD framework, but some are more general. The `.c.src` templating that numpy source code has used for a long time is truly awful, and if all we did was translate that to C++ which was "C plus standard C++ templating", that alone would be a significant improvement.

I'd say that this is true pretty much anywhere one wants to support multiple data types in C.

lorentzenchr commented on doc/neps/roadmap.rst:
```diff
@@ -68,69 +70,119 @@ documentation on many topics are missing or outdated. See :ref:`NEP44` for
+required for a platform to move to a higher tier of support, similar to
+`PEP 11 <https://peps.python.org/pep-0011/>`__.
+
+CPython 3.13 will be the first release to offer a free-threaded build (i.e.,
```
I would promote this to its own section with a title.

Isn‘t it also linked to the above point
> A better story around parallel execution.

lorentzenchr commented on doc/neps/roadmap.rst:
``````````````````````diff
@@ -68,69 +70,119 @@ documentation on many topics are missing or outdated. See :ref:`NEP44` for
+`````````````````````
+The number of downloads of NumPy from PyPI and other platforms continues to
+increase - as of May 2024 we're at >200 million downloads/month from PyPI
+alone). Reducing the size of an installed NumPy package has many benefits:
``````````````````````
```suggestion
alone. Reducing the size of an installed NumPy package has many benefits:
```

lorentzenchr commented on doc/neps/roadmap.rst:
```diff
@@ -68,69 +70,119 @@ documentation on many topics are missing or outdated. See :ref:`NEP44` for
+increase - as of May 2024 we're at >200 million downloads/month from PyPI
+alone). Reducing the size of an installed NumPy package has many benefits:
+faster installs, lower disk space usage, smaller load on PyPI, less
+environmental impact, easier to fit more packages on top of NumPy into an AWS
```
I would reformulate a bit less company specific, e.g
> for resource constraint environment/platforms like AWS Lambda

mhvk commented on doc/neps/roadmap.rst:
```diff
@@ -68,69 +70,119 @@ documentation on many topics are missing or outdated. See :ref:`NEP44` for
+improve the dtype system - see for example :ref:`NEP41` and related NEPs linked
+from it. In NumPy 2.0, a `new C API for user-defined dtypes <https://numpy.org/devdocs/reference/c-api/array.html#custom-data-types>`__
+was made public. We aim
+to encourage its usage and improve this API further.
```
Maybe add ", including support for writing a dtype in Python."? I think that part of the old list is still relevant (and would allow experimenting with dtype metadata perhaps).

rgommers replied:
Good point, that progressed for a bit and then I'm not sure what happened to it. Edit sounds good to me if it's still in someone's plan to do this. @seberg can you comment?

mattip commented:
I think we could add a blurb about working toward limited ABI wheels (abi3). It might be a nice project to see how far we can get using only the limited ABI,and if there are any measurable performance implications. The advantage, in addition to making supporting newer versions of Python easier, would be in needing fewer wheels: only one per platform.

rgommers commented:
Good point @mattip, I like it. In the same vein: I should really flesh out the `numpy-hdronly` idea, that will have benefits in the same area - in particular, no need to build numpy in order to build wheels of downstream packages on platforms we don't publish wheels for.

rgommers committed:
DOC: address all comments, add sections for limited C API and header-only support

[skip actions] [skip azp] [skip cirrus]

rgommers commented:
I addressed all comments and added the limited C API and header-only sections.

I marked this ready for review, it'll be good to get this in, as discussed in the last community meeting.

mattip commented:
Thanks @rgommers. Subsequent fixes can be made in follow-on PRs.
